File: app/src/main/java/net/bloople/stories/Book.java
package net.bloople.stories;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

class Book {
    private long _id = -1L;
    private String path;
    private String title;
    private long mtime;
    private long size;
    private long lastOpenedAt;
    private int lastReadPosition;

    static Book findById(Context context, long id) {
        SQLiteDatabase db = DatabaseHelper.instance(context);

        Cursor result = db.rawQuery("SELECT * FROM books WHERE _id=?", new String[] { String.valueOf(id) });
        result.moveToFirst();

        if(result.getCount() > 0) {
            Book book = new Book(result);
            result.close();
            return book;
        }
        else {
            return null;
        }
    }

    static Book findByPath(Context context, String path) {
        SQLiteDatabase db = DatabaseHelper.instance(context);

        Cursor result = db.rawQuery("SELECT * FROM books WHERE path=?", new String[] { path });
        result.moveToFirst();

        if(result.getCount() > 0) {
            Book book = new Book(result);
            result.close();
            return book;
        }
        else {
            return null;
        }
    }

    Book() {
    }

    Book(Cursor result) {
        _id = result.getLong(result.getColumnIndex("_id"));
        path = result.getString(result.getColumnIndex("path"));
        title = result.getString(result.getColumnIndex("title"));
        mtime = result.getLong(result.getColumnIndex("mtime"));
        size = result.getLong(result.getColumnIndex("size"));
        lastOpenedAt = result.getLong(result.getColumnIndex("last_opened_at"));
        lastReadPosition = result.getInt(result.getColumnIndex("last_read_position"));
    }

    public String path() {
        return path;
    }

    public void path(String path) {
        this.path = path;
    }

    public String title() {
        return title;
    }

    public void title(String title) {
        this.title = title;
    }

    public long mtime() {
        return mtime;
    }

    public void mtime(long mtime) {
        this.mtime = mtime;
    }

    public long size() {
        return size;
    }

    public void size(long size) {
        this.size = size;
    }

    public long lastOpenedAt() {
        return lastOpenedAt;
    }

    public void lastOpenedAt(long lastOpenedAt) {
        this.lastOpenedAt = lastOpenedAt;
    }

    public int lastReadPosition() {
        return lastReadPosition;
    }

    public void lastReadPosition(int lastReadPosition) {
        this.lastReadPosition = lastReadPosition;
    }

    public void save(Context context) {
        ContentValues values = new ContentValues();
        values.put("path", path);
        values.put("title", title);
        values.put("mtime", mtime);
        values.put("size", size);
        values.put("last_opened_at", lastOpenedAt);
        values.put("last_read_position", lastReadPosition);

        SQLiteDatabase db = DatabaseHelper.instance(context);

        if(_id == -1L) {
            _id = db.insert("books", null, values);
        }
        else {
            db.update("books", values, "_id=?", new String[] { String.valueOf(_id) });
        }
    }
}

File: app/src/main/java/net/bloople/stories/Book.kt
package net.bloople.stories

import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.database.Cursor

internal class Book {
    var _id: Long? = null
    var path: String? = null
    var title: String? = null
    var mtime: Long = 0
    var size: Long = 0
    var lastOpenedAt: Long = 0
    var lastReadPosition = 0
    var starred = false
    var openedCount = 0

    constructor()
    constructor(result: Cursor) {
        _id = result.getLong(result.getColumnIndex("_id"))
        path = result.getString(result.getColumnIndex("path"))
        title = result.getString(result.getColumnIndex("title"))
        mtime = result.getLong(result.getColumnIndex("mtime"))
        size = result.getLong(result.getColumnIndex("size"))
        lastOpenedAt = result.getLong(result.getColumnIndex("last_opened_at"))
        lastReadPosition = result.getInt(result.getColumnIndex("last_read_position"))
        starred = result.getInt(result.getColumnIndex("starred")) == 1
        openedCount = result.getInt(result.getColumnIndex("opened_count"))
    }

    fun save(context: Context) {
        val values = ContentValues().apply {
            put("path", path)
            put("title", title)
            put("mtime", mtime)
            put("size", size)
            put("last_opened_at", lastOpenedAt)
            put("last_read_position", lastReadPosition)
            put("starred", if(starred) 1 else 0)
            put("opened_count", openedCount)
        }

        val db = DatabaseHelper.instance(context)
        if (_id != null) {
            db.update("books", values, "_id=?", arrayOf(_id.toString()))
        }
        else {
            _id = db.insert("books", null, values)
        }
    }

    inline fun <R> edit(context: Context, block: Book.() -> R): R {
        return block(this).also { save(context) }
    }

    fun destroy(context: Context) {
        val db = DatabaseHelper.instance(context)
        db.delete("books", "_id=?", arrayOf(_id.toString()))
    }

    fun idTo(intent: Intent): Intent {
        return intent.apply { putExtra("_id", _id) }
    }

    companion object {
        inline fun <R> edit(context: Context, id: Long, block: Book.() -> R): R {
            val book = find(context, id)
            return block(book).also { book.save(context) }
        }

        inline fun <R> editOrNull(context: Context, id: Long, block: Book.() -> R?): R? {
            val book = findOrNull(context, id) ?: return null
            return block(book).also { book.save(context) }
        }

        @JvmStatic
        fun find(context: Context, id: Long): Book {
            val db = DatabaseHelper.instance(context)
            db.rawQuery("SELECT * FROM books WHERE _id=?", arrayOf(id.toString())).use {
                it.moveToFirst()
                return if (it.count > 0) Book(it) else throw NoSuchElementException("Book with id $id not found")
            }
        }

        fun findOrNull(context: Context, id: Long): Book? {
            return try {
                find(context, id)
            } catch(e: NoSuchElementException) {
                null
            }
        }

        @JvmStatic
        fun findByPathOrNull(context: Context, path: String): Book? {
            val db = DatabaseHelper.instance(context)
            db.rawQuery("SELECT * FROM books WHERE path=?", arrayOf(path)).use {
                it.moveToFirst()
                return if (it.count > 0) Book(it) else null
            }
        }

        fun idFrom(intent: Intent?): Long {
            return intent!!.getLongExtra("_id", -1).also {
                if(it == -1L) throw IllegalArgumentException("Intent missing extra _id of type long")
            }
        }

        @JvmStatic
        fun idTo(intent: Intent, id: Long): Intent {
            return intent.apply { putExtra("_id", id) }
        }
    }
}